File: examples/adaptive-ranking-ts/src/adaptive_demo.ts
import { type ExecutableTool, IntentGraph, ToolCatalog } from "@ratel-ai/sdk";

const TOOLS: [string, string][] = [
  ["docker_build", "Build a Docker image from a Dockerfile"],
  ["gh_run_list", "List CI workflow runs and whether the build passed"],
];
const query = "why is the build broken";

const catalog = new ToolCatalog({ method: "semantic" }); // define the tool catalog

await catalog.register(TOOLS.map(([id, description]): ExecutableTool => ({ id, name: id, description, inputSchema: {}, outputSchema: {}, execute: async () => "ok" }))); // tool registration

const graph = new IntentGraph(); // define the intent graph

catalog.enableAdaptiveRanking(graph); // attach it: learn from usage and boost with it

const hits = await catalog.searchAsync(query, 5, "direct", "semantic"); // semantic search, top-5 (dense is async)

hits[0].rank; // 0-based position — order on this, not score
hits[0].fused; // true once the usage arm boosted the result

await catalog.invoke("gh_run_list", {}); // invoke a tool: search + invoke = one observation

graph.clusterCount; // clusters learned
graph.rev; // write counter — persist only when it changes

const saved = graph.toJson(); // serialize the in-memory graph
const graph2 = IntentGraph.fromJson(saved); // reload it (invalid graphs are rejected)

catalog.adaptiveRankingStatus.status; // active | inactive | unknown | paused: model mismatch

await catalog.rebuildIntentGraph(); // re-embed under the current model (recover after a model swap)

catalog.enableAdaptiveRanking(graph2, { rebuildOnModelChange: true }); // default false; true auto-recovers on next search

catalog.disableAdaptiveRanking(); // turn off; the graph keeps what it learned

File: examples/adaptive-ranking-python/adaptive_demo.py
from ratel_ai import ExecutableTool, IntentGraph, ToolCatalog

TOOLS = [
    ("docker_build", "Build a Docker image from a Dockerfile"),
    ("gh_run_list", "List CI workflow runs and whether the build passed"),
]
query = "why is the build broken"

catalog = ToolCatalog(method="semantic")  # define the tool catalog

await catalog.register([ExecutableTool(id=i, name=i, description=d, execute=lambda _a: "ok") for i, d in TOOLS])  # tool registration

graph = IntentGraph()  # define the intent graph

catalog.enable_adaptive_ranking(graph)  # attach it: learn from usage and boost with it

hits = await catalog.search_async(query, 5, method="semantic")  # semantic search, top-5 (dense is async)

hits[0].rank   # 0-based position — order on this, not score
hits[0].fused  # True once the usage arm boosted the result

await catalog.invoke("gh_run_list", {})  # invoke a tool: search + invoke = one observation

graph.cluster_count  # clusters learned
graph.rev            # write counter — persist only when it changes

saved = graph.to_json()               # serialize the in-memory graph
graph = IntentGraph.from_json(saved)  # reload it (invalid graphs are rejected)

catalog.adaptive_ranking_status  # active | inactive | unknown | paused: model mismatch

await catalog.rebuild_intent_graph()  # re-embed under the current model (recover after a model swap)

catalog.enable_adaptive_ranking(graph, rebuild_on_model_change=True)  # default False; True auto-recovers on next search

catalog.disable_adaptive_ranking()  # turn off; the graph keeps what it learned
